Add rrmscorer

Imanjdd opened the pull request:
FOR CONTRIBUTOR:
* [ ] I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ ] License permits unrestricted use (educational + commercial)
* [ ] This PR adds a new tool or tool collection
* [ ] This PR updates an existing tool or tool collection
* [ ] This PR does something else (explain below)

### **Abstract**

This PR introduces “RRM-Scorer,” a tool designed to predict RNA Recognition Motif (RRM) scores for protein sequences based exclusively on their sequences, enabling predictions for interactions with any RNA target. This functionality is essential for researchers focusing on RNA-binding proteins and provides an effective, sequence-only method to assess binding affinities without the need for experimental interaction data.

Description of the new tool:

RRM-Scorer uses the sequence information of proteins to predict how likely they are to bind to RNA via their RNA Recognition Motifs (RRMs). This is achieved through a meticulously crafted alignment of RRM structures in complex with RNA. By analyzing interaction patterns within these complexes, scores are derived that quantify binding propensities.

About the Bio2Byte research group
We investigate how the dynamics, conformational states, and available experimental data of proteins relate to their amino acid sequence. Underlying physical and chemical principles are computationally unraveled through data integration, analysis, and machine learning, so connecting them to biological events and improving our understanding of the way proteins work.



### Main features:

- Supports input via FASTA formatted sequences or UniProt identifiers.

- Users can specify RNA targets for more tailored predictions.

- Provides options to generate graphical plots of the predictions and alignments of the sequences as output.

- Utilizes a window-based analysis for predictions, enhancing the tool’s flexibility and applicability to various research scenarios.

Examples of Usage

This section provides clear examples of how to run the RRM-Scorer tool using different options. The examples aim to demonstrate typical uses and help new users get started with the tool.

Running RRM-Scorer on a Protein Identifier with RNA Target Sequence

To predict RNA Recognition Motif (RRM) scores for the protein with UniProt ID P19339 and a specific RNA sequence, you can use the following command:
`$ python -m rrmscorer -u P19339 -r AUGGCU -w 5 -j output/ -c output/ `

Related Links:

-  PyPi Package: [RRM-Scorer on PyPi](https://pypi.org/project/rrmscorer/)

- Tool Homepage: [RRM-Scorer at Bio2Byte](https://bio2byte.be/rrmscorer/)

- Bio2Byte Main Website: [Bio2Byte](https://bio2byte.be/)


### **Examples** 

Here are examples of outputs based on the previous command:

Json format : 

sp_P19339_SXL_DROME_209-292 :
`
{
  "AUGGC": -0.7819223678574603,
  "UGGCU": -0.8964132211484258
}
`
sp_P19339_SXL_DROME_119-206 : 
`
{
  "AUGGC": -0.5683559409751302,
  "UGGCU": -1.0775178980839322
}
`



Tabular format : 

sp_P19339_SXL_DROME_209-292 :
`
AUGGC,-0.7819223678574603
UGGCU,-0.8964132211484258
`


sp_P19339_SXL_DROME_119-206 : 


`
AUGGC,-0.5683559409751302
UGGCU,-1.0775178980839322
`

### **Test Cases**

Here are the detailed descriptions for each test case of the RRM-Scorer tool :

### Test Descriptions for RRM-Scorer

#### Test 1 - Basic Functionality with UniProt ID and RNA target
- **Objective**: To validate that the tool handles basic inputs, including a UniProt ID and a specified RNA sequence, correctly.
- **Expectation**: The tool should generate expected outputs in JSON and tabular formats without errors, using the specified UniProt ID and RNA sequence.
- **Result**: Pass if the tool correctly processes the inputs and generates outputs with the correct options specified in the command.

#### Test 2 - Full Input with Outputs Verification
- **Objective**: To test the tool's ability to process a UniProt ID with an RNA target and generate multiple output types.
- **Expectation**: All specified outputs (CSV, JSON, but no FASTA or plots) should be accurately generated based on the provided inputs.
- **Result**: Pass if all expected output files are generated in the correct formats and counts.

#### Test 3 - Handling Multiple Outputs without Plot
- **Objective**: To verify the tool's performance when multiple output formats are requested without generating plots.
- **Expectation**: The tool should successfully generate CSV and JSON outputs based on RNA target and UniProt ID without creating plot files.
- **Result**: Pass if the outputs are correctly generated and no plot files are produced.

#### Test 4 - Overload with Multiple Options (Intended to Fail)
- **Objective**: To assess how the tool manages an overload by executing multiple options simultaneously and to verify correct output count.
- **Expectation**: The tool is expected to either handle the multiple options incorrectly or mismanage the output files. Specifically, the 'split_csv' output collection is expected to contain 123 elements, but an error is anticipated to lead to a different count.
- **Result**: Expected to fail, demonstrating the tool's error handling capacity when multiple options are mismanaged and output collections do not meet the expected counts.

#### Test 5 - Basic RNA Target Specification
- **Objective**: To confirm that the tool processes commands with just basic RNA targeting without additional options efficiently.
- **Expectation**: Correct processing of RNA sequence and UniProt ID without additional outputs like plots or FASTA.
- **Result**: Pass if the tool executes without errors and generates only the specified outputs.

#### Test 6 - Generate FASTA Files with RNA Target
- **Objective**: To test the FASTA file generation functionality with RNA targeting.
- **Expectation**: The tool should generate FASTA files alongside JSON and tabular data.
- **Result**: Pass if FASTA files are correctly generated along with other specified output formats.

#### Test 7 - Top Scoring RNA without Custom Target
- **Objective**: To evaluate the tool’s ability to handle top RNA scoring without a custom RNA sequence.
- **Expectation**: The tool should process inputs and generate outputs without RNA sequences, focusing solely on protein data.
- **Result**: Pass if outputs are correctly generated and aligned with input settings, particularly not requiring an RNA sequence.

#### Test 8 - Complex Command with Multiple Outputs and Plotting
- **Objective**: To verify the tool's ability to manage complex input including UniProt ID, RNA targeting, and multiple output formats with plotting.
- **Expectation**: The tool should handle multiple outputs and generate plots correctly.
- **Result**: Pass if all outputs, including plots, are correctly generated and the tool handles complex inputs seamlessly.

#### Test 9 - Plot Generation without RNA Target (Expected to Fail)
- **Objective**: To examine the tool’s error handling when plotting is requested without an RNA target.
- **Expectation**: An error should occur, indicating that RNA input is necessary for plotting.
- **Result**: Expected to fail, confirming the tool's requirement for RNA input when plotting.

#### Test 10 - Plotting without Specific Predictors
- **Objective**: To test the tool’s plotting functionality when no specific RNA predictors are selected.
- **Expectation**: The tool should manage the absence of specific predictors gracefully and not generate plots.
- **Result**: Pass if no plots are generated and the tool does not fail unexpectedly.

#### Test 11 - Comprehensive Input with All Features
- **Objective**: To test all functionalities including UniProt ID, custom RNA target, window size adjustments, and generation of all output types.
- **Expectation**: The tool should handle all inputs correctly and generate all outputs as specified.
- **Result**: Pass if the tool functions as expected and all outputs, including FASTA files and CSV, are generated.

#### Test 12 - Handling Incorrect Window Size (Expected to Fail)
- **Objective**: To verify the tool's response to an invalid window size input.
- **Expectation**: The tool should report an error due to the invalid window size.
- **Result**: Expected to fail, confirming that the tool properly validates the window size input.

#### Test 13 - Multiple Outputs with Advanced Options
- **Objective**: To evaluate the tool's capability to process multiple advanced options, including plotting and FASTA generation.
- **Expectation**: The tool should correctly handle and execute the complex command with all specified outputs.
- **Result**: Pass if the execution is successful and all outputs are correctly generated as per the command options.

#### Test 14 - Fasta Input with Custom RNA Target and Output Verification (Expected to Fail)
- **Objective**: To specifically test the Fasta input functionality and verify the generation of output files.
- **Expectation**: An error indicating that a UniProt ID should not be used when a Fasta input is provided.
- **Result**: Expected to fail, highlighting that Fasta inputs are not to be mixed with UniProt IDs in the command line.

#### Test 15 - Window Size Validation (Expected to Fail)
- **Objective**: To test the tool's validation of the `window_size` parameter when an unsupported size is provided.
- **Expectation**: The tool should validate the `window_size` parameter against the accepted values (3 and 5), and fail when an invalid option is used.
- **Result**: Expected to fail, confirming that the tool correctly enforces the restrictions on the window size parameter.

### **Screenshots**

Tool configuration:
<img width="845" alt="Capture d’écran 2024-07-08 à 03 23 36" src="https://github.com/galaxyproject/tools-iuc/assets/167193137/3b2a8907-bdf9-4fdf-a627-e137072d2b79">

<img width="845" alt="Capture d’écran 2024-07-08 à 03 24 23" src="https://github.com/galaxyproject/tools-iuc/assets/167193137/d6ccd9ac-3345-49dc-b5da-43f793c1d8c1">

Tool results in Json formats :
<img width="1126" alt="Capture d’écran 2024-07-08 à 03 26 56" src="https://github.com/galaxyproject/tools-iuc/assets/167193137/87e206c9-d3d0-4241-8d4d-3de3ffb353fa">

Tool results in tabular format :

<img width="1136" alt="Capture d’écran 2024-07-08 à 03 27 23" src="https://github.com/galaxyproject/tools-iuc/assets/167193137/b4e1823d-fe73-4667-85d1-330150fb16c1">

Tool results in Fasta format : 

<img width="1123" alt="Capture d’écran 2024-07-08 à 03 29 53" src="https://github.com/galaxyproject/tools-iuc/assets/167193137/e39268b6-253c-4d9e-9282-be5e5722f8b6">

bgruening commented on tools/rrmscorer/.shed.yml:
```diff
@@ -0,0 +1,17 @@
+long_description: |
+    RRMScorer allows the user to predict RNA recognition motif binding scores solely from sequence data. More details can be found on the [Bio2Byte RRMScorer webpage](https://bio2byte.be/rrmscorer/).
+homepage_url: https://bio2byte.be/rrmscorer/
+remote_repository_url: https://github.com/Bio2Byte/tools-iuc
```
this should be the IUC main repo, not a fork

Imanjdd replied:
Okay, thanks! I will make that change.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+<tool id="rrmcorer" name="RRM-Scorer: quick predictions RRM-RNA binding" version="1.0.9">
```
Can you please use a macro for the version and follow the best practices here: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+<tool id="rrmcorer" name="RRM-Scorer: quick predictions RRM-RNA binding" version="1.0.9">
+    <description>Predict RNA Recognition Motif (RRM) scores for protein sequences and any RNA target purely based on their sequences.</description>
```
the description is a bit too long I think, can this be sharpend?

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+    </requirements>
+   <command detect_errors="exit_code"><![CDATA[
+mkdir -p ./json ./tabular ./fasta ./plots &&
+#set $options = []
```
Its a bit unusual to use the options array. Any reason you need to do that?

Imanjdd replied:
No specific reason; it was just an initial choice. However, now I’ve made some changes.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+                    <param name="target" type="text" label="Target RNA sequence (min. 5 nucleotides)" help="Provide a valid target RNA sequence."/>
+                </when>
+            </conditional>
+            <param name="window_size" type="select" label="The window size to test (Only 3 and 5 nucleotide windows are accepted)">
```
```suggestion
            <param name="window_size" type="select" label="The window size to test">
```

Not needed in this interface.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+        </section>
+
+        <section name="tool_parameters" title="Tool parameters" help="Configure this section to select the predictions to be executed">
+            <conditional name="custom_rna_target">
```
Do you need this conditional at all. You could just make the text param optional isn't it?

Imanjdd replied:
Yes, I can of course.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+            <conditional name="custom_rna_target">
+                <param name="custom_rna_target" type="boolean" label="Custom RNA target" help="Enable custom RNA target for the predictions." />
+                <when value="true">
+                    <param name="target" type="text" label="Target RNA sequence (min. 5 nucleotides)" help="Provide a valid target RNA sequence."/>
```
you can add a validator to check that the string is only RNA

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+        <output_collection name="split_csv" type="list" count="2"/>
+        <output_collection name="split_json" type="list" count="2"/>
+        <output_collection name="split_fasta" type="list" count="0"/>
+        <output_collection name="split_plots" type="list" count="0"/>
```
Does it make sense to create collections with 0 elements? Do you know beforehand when a collection will be empty? If so you can not create the collection at all.

Imanjdd replied:
I created collections with 0 elements intentionally to test and ensure that all outputs, even those expected to be empty, were functioning correctly. Is this approach helpful, or should I avoid creating empty collections ?

Imanjdd replied:
Yes, I understand now. Thanks for the clarification—I’ll make sure to avoid creating empty collections.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,411 @@
+        </assert_command>
+
+        <output_collection name="split_csv" type="list" count = "2"/>
+        <output_collection name="split_json" type="list" count = "2"/>
```
very exhausive tests, thanks a lot! Can you maybe also test the collection content? E.g. check that its a valid json etc or found the columns of a CSV.

https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output-assert-contents

bgruening reviewed (commented):
Thanks a lot @Imanjdd! Also for the very cool PR description.

I have added some comments inline.

Imanjdd commented:
> Thanks a lot @Imanjdd! Also for the very cool PR description.
> 
> I have added some comments inline.

Thanks a lot for the feedback! Your comments are really helpful for making improvements. I’ll start working on those right away.

> Thanks a lot @Imanjdd! Also for the very cool PR description.
> 
> I have added some comments inline.

Thanks a lot for the feedback @bgruening ! Your comments are really helpful for making improvements. 
I’ll start working on those right away.

Imanjdd commented:
> Thanks a lot @Imanjdd! Also for the very cool PR description.
> 
> I have added some comments inline.

Thanks a lot for the feedback @bgruening ! Your comments are really helpful for making improvements. 
I’ll start working on those right away.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,402 @@
+<tests>
+   
+
+    <test>
```
Try adding here `expect_num_output` https://docs.galaxyproject.org/en/latest/dev/schema.html#id99

bgruening replied:
ping

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,402 @@
+    
+    <citations>
+        <citation type="bibtex">@misc{rrmscorer, title={RRMScorer on PyPI}, howpublished={\url{https://pypi.org/project/rrmscorer/}}}</citation>
+        <citation type="bibtex">@misc{bio2byte, title={Bio2Byte RRMScorer}, howpublished={\url{https://bio2byte.be/rrmscorer}}</citation>
```
Is there a DOI that we can cite?

Imanjdd replied:
Yes, I will add the DOI

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,402 @@
+    #if $plot_options.generate_fasta
+        -a 'fasta'
+    #end if
+    && ls -l ./json ./tabular ./fasta ./plots
```
this is not needed anymore I guess

bgruening replied:
```suggestion
```

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,402 @@
+
+        <section name="plot_options" title="Plot options" help="Configure plot output">
+            <param name="generate_plots" type="boolean" label="Generate score plots for all the RNA possible windows" help="Enable to generate score plots." />
+            <param name="top_scoring_rna" type="boolean" label="Find the top-scoring RNA for the specified RRM(s) and plot their sequence logo" help="Enable to find and plot the top-scoring RNA." />
```
```suggestion
            <param name="top_scoring_rna" type="boolean" truevalue="-t" falsevalue="" label="Find the top-scoring RNA for the specified RRM(s) and plot their sequence logo" help="Enable to find and plot the top-scoring RNA." />
```

This is a nice trick you can use ...

Imanjdd replied:
thank you !

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,402 @@
+    #if $tool_parameters.target
+        -r '$tool_parameters.target'
+    #end if
+    #if $plot_options.top_scoring_rna
+        -t
+    #end if
```
```suggestion
    $plot_options.top_scoring_rna
```

With the trick below, booleans become like this ... in the falsevalue case its just an empty string. In the truevalue case its `-t`

bgruening replied:
ping

Imanjdd replied:
I had made the modification, but I ran into some errors, so I’ve temporarily reverted the changes for now

bgruening replied:
please let me know the errors or ping me. It should be changed into the simpler version imho.

Imanjdd replied:
I’ve just made the modification and it’s working now! Would you like to take a look at the code section?

bgruening commented:
If you like please also look at https://github.com/galaxyproject/galaxy-language-server it will help you with the linting etc ...

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,385 @@
+<tool id="rrmcorer" name="RRM-Scorer: quick predictions RRM-RNA binding" version="@TOOL_VERSION@" profile="22.05">
```
```suggestion
<tool id="rrmcorer" name="RRM-Scorer" version="@TOOL_VERSION@" profile="22.05">
    <description>predictions of RRM-RNA binding</description>
```

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,385 @@
+    && ls -l ./json ./tabular ./fasta ./plots
+]]></command>
+    <inputs>
+        <section name="input_sequence" title="Input sequence">
```
```suggestion
        <section name="input_sequence" title="Input sequence" expanded="true">
```

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,385 @@
+            <has_text text= "-c 'tabular'"/>
+            <has_text text= "-p 'plots'"/>
+
+        </assert_command>
+
+        
+
+        </test>
+
+    
```
```suggestion
            </assert_command>
        </test>
```

A bit of cleanup would be nice.

agdiaz commented:
Hello @bgruening, this is @agdiaz from Bio2Byte group. Thanks for your review. @Imanjdd joined us for an internship to contribute to deploying our tools on Galaxy from May to July 24'. I'm glad to complete soon the pending comments and help to integrate her contributions to this repository. Soon we'll see her work on Galaxy Europe :)

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,379 @@
+    </command>
+    <inputs>
+        <section name="input_sequence" title="Input sequence" expanded="true">
+            <param name="fasta_input" type="data" format="fasta" optional="true" label="Protein sequence(s) FASTA format" help="Provide a FASTA file containing the protein sequences or leave blank if specifying a UniProt ID."/>
```
You should model that with a conditional I think.

This way the user needs to put one option in and you will never have the else condition above where you raise an error on the cluster. This will safe compute resources.

bgruening commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,379 @@
+    <inputs>
+        <section name="input_sequence" title="Input sequence" expanded="true">
+            <param name="fasta_input" type="data" format="fasta" optional="true" label="Protein sequence(s) FASTA format" help="Provide a FASTA file containing the protein sequences or leave blank if specifying a UniProt ID."/>
+            <param name="uniprot_id" type="text" value="P19339" optional="true" label="OR Protein identifier from Uniprot" help="Provide a UniProt ID (e.g. P19339) or leave blank if specifying a FASTA file.">
```
the help will be easier if you use a conditional here.

bgruening reviewed (commented):
Two remaining comments :)

Thanks a lot @agdiaz!

SaimMomin12 reviewed (commented):
Some preliminary comments inline

agdiaz commented:
Hi @SaimMomin12 and @bgruening, all the comments have been addressed. The CI returned green. Would it be possible to ask for another review round ? Thanks a lot for the feedback, the final state of this tool is amazing, so proud of the results. We'll continue after RRMScorer with our next in-house tool, Constava, in the upcoming days.

SaimMomin12 commented on tools/rrmscorer/RRMScorer.xml:
```diff
@@ -0,0 +1,515 @@
+<tool id="rrmscorer" name="RRM-Scorer" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
+    <description>predicts RNA Recognition Motif (RRM) scores for protein sequences.</description>
```
```suggestion
    <description>Predicts RNA Recognition Motif (RRM) scores</description>
```
Can we perhaps make this shorter?

SaimMomin12 reviewed (commented):
Looks good to me, just a small comment :) I think we change then merge. Thanks @agdiaz

agdiaz commented:
Hi @SaimMomin12 , thank you for merging the PR into the main one. I'd like to ask how to proceed in order to make the RRMScorer tool visible in the European Galaxy instance? Will it be automatically done or I need to open a PR in a different repository ? Thanks in advance for your reply. All the best from Belgium 👋
